### PR TITLE
Clean up the demo console interface

### PR DESCRIPTION
**Pull request by mbasaglia**
This is part of the work to improve demos (see Unvanquished/Unvanquished#988).
It does the following:
* Renames cvars to `demo.*`
* Renames commands to `demo_*`
* Removes demo-related functionality that is completely broken (eg: audio recording)
* Fixes CMake to avoid overwriting submodules during builds

### CODE REVIEW

File: src/engine/client/cl_main.cpp [hunk 1]
```diff
@@ -302,213 +307,204 @@ void CL_WriteDemoMessage( msg_t *msg, int headerBytes )
...
+    int len = -1;
+    FS_Write( &len, 4, clc.demofile );
+    FS_Write( &len, 4, clc.demofile );
+    FS_FCloseFile( clc.demofile );
+    clc.demofile = 0;
+
+    clc.demorecording = false;
+    Cvar::SetValueForce(cvar_demo_status_isrecording.Name(), "0");
+    Cvar::SetValueForce(cvar_demo_status_filename.Name(), "");
+    Log::Notice("%s", "Stopped demo.\n" );
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: Is the extra newline intentional? Log:: functions should add a newline by themselves. (If it is, comment so)

File: src/engine/client/cl_main.cpp [hunk 2]
```diff
@@ -302,213 +307,204 @@ void CL_WriteDemoMessage( msg_t *msg, int headerBytes )
...
-
-record <demoname>
-
-Begins recording a demo from the current position
-====================
-*/
-
-static char demoName[ MAX_QPATH ]; // compiler bug workaround
-void CL_Record_f()
+class DemoStopRecordCmd: public Cmd::StaticCmd
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: nit: The command is demo_record_stop but the function is DemoStopRecord. Can the two be consistent?

File: src/engine/client/cl_main.cpp [hunk 3]
```diff
@@ -302,213 +307,204 @@ void CL_WriteDemoMessage( msg_t *msg, int headerBytes )
...
-	if ( clc.demorecording )
-	{
-		Log::Warn("Already recording." );
-		return;
-	}
+    void Run(const Cmd::Args&) const OVERRIDE
+    {
+        if ( !clc.demorecording )
+        {
+            Log::Notice("%s", "Not recording a demo.\n" );
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: ditto comment about newlines.

Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: Also, how come sometimes you do Log::Notice ("%s", "Static Text");, but other times you just do Log::Notice("Static text"); ?

File: src/engine/client/cl_main.cpp [hunk 4]
```diff
@@ -302,213 +307,204 @@ void CL_WriteDemoMessage( msg_t *msg, int headerBytes )
...
 
-	if ( Cmd_Argc() == 2 )
-	{
-		s = Cmd_Argv( 1 );
-		Q_strncpyz( demoName, s, sizeof( demoName ) );
-		Com_sprintf( name, sizeof( name ), "demos/%s.dm_%d", demoName, PROTOCOL_VERSION );
-	}
+    void Run(const Cmd::Args& args) const OVERRIDE
+    {
+        if ( args.size() > 2 )
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: Will this prevent people from just calling demo_record?

File: src/engine/client/cl_main.cpp [hunk 5]
```diff
@@ -302,213 +307,204 @@ void CL_WriteDemoMessage( msg_t *msg, int headerBytes )
...
-		{
-			if ( *period == '/' )
-			{
-				break;
-			}
-		}
+        if ( args.size() == 2 )
+        {
+            demo_name = args[1];
+        }
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: imo, you should generate a name here for readability. It's unclear to the reader that you're going to generate a name if the demo name is empty.

File: src/engine/client/cl_main.cpp [hunk 6]
```diff
@@ -863,31 +678,28 @@ class DemoCmd: public Cmd::StaticCmd {
             return {};
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: There should already be helper functions for command completion that just lists files in a directory and filters on a pattern.

File: src/engine/client/cl_main.cpp [hunk 7]
```diff
@@ -926,15 +738,7 @@ void CL_ShutdownAll()
...
-	if ( clc.demorecording )
-	{
-		CL_StopRecord_f();
-	}
-
-	if ( clc.waverecording )
-	{
-		CL_WavStopRecord_f();
-	}
+    CL_StopRecord();
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: indentation looks wrong... I know we are fighting competing styles, but we should at least match the code that we are around.

File: src/engine/client/cl_parse.cpp [hunk 8]
```diff
@@ -234,43 +234,7 @@ void CL_ParseSnapshot( msg_t *msg )
 		{
 			if ( cl_autorecord->integer /*&& Cvar_VariableValue( "g_synchronousClients") */ )
```
Comment (DolceTriade) [inline src/engine/client/cl_parse.cpp]: Let's delete the commented out code as well

File: src/engine/client/cl_main.cpp [hunk 9]
```diff
@@ -302,213 +307,204 @@ void CL_WriteDemoMessage( msg_t *msg, int headerBytes )
...
-	Log::Notice("%s", "Stopped demo.\n" );
+    // finish up
+    int len = -1;
+    FS_Write( &len, 4, clc.demofile );
+    FS_Write( &len, 4, clc.demofile );
+    FS_FCloseFile( clc.demofile );
+    clc.demofile = 0;
+
+    clc.demorecording = false;
+    Cvar::SetValueForce(cvar_demo_status_isrecording.Name(), "0");
```
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: Why not just use Cvar::Set here?

Comment (mbasaglia) [inline src/engine/client/cl_main.cpp]: Bacause it's a read-only cvar and Set doesn't work on those.

File: src/engine/client/cl_main.cpp (hunk 5 above)
Comment (mbasaglia) [inline src/engine/client/cl_main.cpp]: CL_Record is called elsewhere with an automatically-generated name, it used to have the generation code duplicated so I moved it in CL_Record.

File: src/engine/client/cl_main.cpp (hunk 4 above)
Comment (mbasaglia) [inline src/engine/client/cl_main.cpp]: Nope

File: src/engine/client/cl_main.cpp (hunk 6 above)
Comment (mbasaglia) [inline src/engine/client/cl_main.cpp]: Isn't it already using it?

File: src/engine/client/cl_main.cpp (hunk 9 above)
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: Weird. Didn' t know that.

File: src/engine/client/cl_main.cpp (hunk 5 above)
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: I guess. It's just unclear from reading this that it does that. I don't feel strongly here.

File: src/engine/client/cl_main.cpp (hunk 4 above)
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: Read it wrong. :(

File: src/engine/client/cl_main.cpp (hunk 1 above)
Comment (mbasaglia) [inline src/engine/client/cl_main.cpp]: It was already like that so I don't know

File: src/engine/client/cl_main.cpp (hunk 6 above)
Comment (DolceTriade) [inline src/engine/client/cl_main.cpp]: It is. My bad.

**Comment by Viech** [comment]
Note that stuff like 9c3071a should ideally go in a PR of their own. 😉

**Pull request merged**